### ISSUE TITLE
iOS - "Wrong JPEG library version: library is 90, caller expects 62"

### ISSUE THREAD

**Issue by ilijapuaca** [libjpeg-turbo/libjpeg-turbo#140]
This might not be the right place to post this, but I'm banging my head for 2 days straight trying to figure this one out.

I've built mozjpeg for iOS per instructions provided in BUILDING.txt file. Everything went fine and I wrote some code that should perform the encoding. When trying to retrieve the handle using tjInitCompress, it returns NULL and the error string is the one from the title.

I'm not that knowledgable when it comes to exact details about how building/linking works for this particular library, but as far as I've figured it builds libjpeg from source code that is part of the repo. Does it for whatever reason have dynamic link to system libjpeg that would use the one that's available on iOS?

I understand that there are no plans for turbojpeg to support jpeg9 for somewhat obvious reasons, with that in mind, should it be possible to get mozjpeg to work with iOS10/XCode8 setup?

**Comment by dcommander** [comment]
That error would only occur if you built your application against a libjpeg-turbo dynamic library, but somehow the application was loading libjpeg v9 at run time.  No idea why that would occur, since I wasn't aware that a libjpeg dynamic library existed in iOS.  As far as I know, Apple doesn't use libjpeg or libjpeg-turbo for handling JPEG images, or if they do, it's hidden in some other library framework.

Also, why are you using mozjpeg for an iOS app?  mozjpeg is designed for web developers who want to compress JPEGs using the maximum possible compression, but the tradeoff is extreme-- you may get 15 or 20% better compression, but it uses something like 40-50x more CPU time than libjpeg-turbo.

mozjpeg is a downstream project with vastly different goals than ours.  It is not supported by me or other libjpeg-turbo project members, unless there is an issue that affects both mozjpeg and libjpeg-turbo.

The latest binaries of libjpeg-turbo provide iOS forks for ARMv7, ARMv7s, and ARMv8 with bitcode enabled.  The easiest way to use libjpeg-turbo in an iOS app is to download the libjpeg-turbo DMG file, install it on your Mac, and link your app against /opt/libjpeg-turbo/lib/libturbojpeg.a (TurboJPEG API) or /opt/libjpeg-turbo/lib/libjpeg.a (libjpeg API), both of which contain the necessary iOS architectures to support iOS 6.0 and later on all iPads and all iPhones newer than the 3G.

If you still want or need to build from source for some reason, then try building libjpeg-turbo rather than mozjpeg.  If you are still encountering the error, then I will work with you to diagnose it, but I can't provide any tech support for downstream projects such as mozjpeg.

**Comment by ilijapuaca** [comment]
I eventually got it working by linking with both libjpeg.a **and** libjpeg-turbo.a that were the result of the build. I assumed it was only libjpeg-turbo.a that I needed and that's the one I created a .framework file out of.

> Also, why are you using mozjpeg for an iOS app? mozjpeg is designed for web developers who want to compress JPEGs using the maximum possible compression, but the tradeoff is extreme-- you may get 15 or 20% better compression, but it uses something like 40-50x more CPU time than libjpeg-turbo.

In the app I'm currently working on, the biggest issue we have is the amount of data (images) that we upload to our servers. We were trying to compare different options we have for compression and pick the one that suits us the best. After running some tests with webp, we wanted to try out mozjpeg.

As you suggested, the CPU load doesn't seem to be worth the file size decrease (surprisingly it did not go down as much as we expected).

Either way, all of what you wrote above makes sense, and the issue is obviously not with neither libjpeg-turbo nor mozjpeg. I still don't quite get what libjpeg was being linked against when I was getting the error initially, but I'm closing the issue down.

**Comment by dcommander** [comment]
That's odd.  libturbojpeg.a should be self-contained.  For instance:

```
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang -arch arm64 -miphoneos-version-min=7.0 -fembed-bitcode -funwind-tables -O3 -isysroot  /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS.sdk tjunittest.c -o tjunittest /opt/libjpeg-turbo/lib/libturbojpeg.a 
```

works fine without linking with libjpeg.a.

**Comment by ilijapuaca** [comment]
I was doing this:

`configure --host  aarch64-apple-darwin --enable-shared=false CC="/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang" LD="/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang" CFLAGS="-mfloat-abi=softfp -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS.sdk -O3 -arch arm64 -miphoneos-version-min=7" LDFLAGS="-mfloat-abi=softfp -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS.sdk -arch arm64 -miphoneos-version-min=7" CCASFLAGS="-no-integrated-as -arch arm64 -miphoneos-version-min=7"`

`
make -j 4
`

This would create two static libraries, libturbojpeg.a and libjpeg.a. Since I wanted to use libjpeg-turbo API, I used libturbojpeg.a and created a .framework out of it. Trying to use the API produced the error I mentioned in the title.

After I inserted libjpeg.a in my project and linked against it, the error was gone. Don't know if that sheds any light on the situation.

**Comment by dcommander** [comment]
It doesn't, unfortunately.  I tried building libjpeg-turbo using the identical command line that you used above, then I did:

``/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang -mfloat-abi=softfp -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS.sdk -O3 -arch arm64 -miphoneos-version-min=7 -o tjunittest.static ../tjunittest.c .libs/libturbojpeg.a
``

Works beautifully and does not require linking with libjpeg.  I am guessing that the problem was introduced when you created a framework out of it.
